### DISCUSSION TITLE
Generate Layer with Countries names

### DISCUSSION THREAD

**Discussion by BenOnTrack** [systemed/tilemaker#756] [Q&A]
I'm trying to use the ne_10m_admin_0_countries from Natural Earth Data, which is supposed to contain countries name (and ISO) according to the .dbf file.

I've added `"source_columns":true` in my tilemaker file .config while my tilemaker .process is

```
function init_function()
end
function exit_function()
end

node_keys = {}
function node_function()
end

function way_function()
end

function attribute_function(attr,layer)
	return attr
end
```

and then create a style in my maplibre-gl style
```
 {
      "id": "countries",
      "type": "symbol",
      "source": "countries",
      "source-layer": "countries",
      "metadata": {
        "switch": false,
        "clickable": true
      },
      "layout": {
        "text-field": "{ADMIN} {ISO_A2} {ISO_A3}",
        "text-font": [
          "Noto Sans Italic"
        ],
        "text-max-width": 6.25,
        "visibility": "visible"
      },
      "paint": {
        "text-color": "#334",
        "text-halo-blur": 1,
        "text-halo-color": "rgba(255,255,255,0.8)",
        "text-halo-width": 2
      }
    }
```
but got nothing rendered... any hints ?

**Comment by cldellow** [comment]
Step 1: Is tilemaker emitting the data?

If you are unsure, you can:

1. output a .pmtiles file
2. open [PMTiles Viewer](https://pmtiles.io/) and point it at your local PMTiles file
3. zoom in to the zoom level where you expect to see it - maybe disable other layers in the viewer to make it more obvious
4. enable `show attributes` option and hover over the tile

If you don't see the object in step 3, then tilemaker is not emitting _anything_. Put a `print(...)` in your `attribute_function` to see if it's even getting called.

If you see the object in step 3, but don't see the expected attributes in step 4, then tilemaker's emitting the shape, but not the attributes.

If either case applies, further debugging will require that you also include the relevant portion of your `config.json` -- the `layers` entry and the overall `minzoom` / `maxzoom` values.

**Comment by BenOnTrack** [comment]
Thank you for your quick feedback.
I've been working with .mbtiles and can't find any 'debugging' tool that would be able to analyse the file without the need to generate a style.json.

I'll then try to output a .pmtiles to check if the data is actually output.

I'm not entirely sure about the .process file if I need to code anything in the 'node_function' but it doesn't look like I should since I'm requesting all tables from the initial .shp to be printed out.

It might as well be that the natural earth data I'm using is actually not including the name of the countries (but then it would mean that their description of the content of the file is wrong though...)

**Comment by cldellow** [comment]
Yes, the process.lua file looks fine. If there's an issue it'll be in the config.json or the style file.

This is what I see when I use the pmtiles inspector on that NE data, just passing through the data from the shapefile untransformed, as you are trying to do:

![Selection_773](https://github.com/user-attachments/assets/376fd075-9bce-4e9e-872d-350ef08ef56a)

Annoyingly, the inspector doesn't seem to sort the attributes, but `ADMIN`, `ISO_A2` and `ISO_A3` are all there.

**Reply by BenOnTrack** [reply]
That's great news. So I know where to focus - thanks, I'll come back to this thread this evening.

**Reply by BenOnTrack** [reply]
So i confirm the data get logged when generating a .pmtiles, so I figured it might be an issue with the style. 
To make sure I would take control of what's happening, I've decided to modify the label of my attribute and changed the .process to `return {name=attr["ADMIN"]}` 
and it works, with a style with `"text-field": "{name}"` instead of `"text-field": "{ADMIN}"`

So I would recommend to modify the README and indicate (if anyone has the answer) where the attributes ends up, but if an attribute `ADMIN` exists in your .shp file, you won't be able to query it with `"text-field": "{ADMIN}"` unless you re-align it with `return {ADMIN=attr["ADMIN"]}`.

Happy to hear other's thoughts

**Reply by cldellow** [reply]
Interesting, it works for me without doing this remapping step.

When you say `I confirm the data get logged when generating a .pmtiles` - what does that mean? Did you see the data in the .pmtiles when using the PMTiles Viewer tool? If so, then tilemaker was working as expected, and the issue was with your style. It would be good to know the details here so we can understand if there was a bug in tilemaker or not.

**Reply by cldellow** [reply]
If you can reproduce this issue, can you please create a gist that contains your process.lua, config.json and style.json files? That will make this much less of a game of telephone.

**Reply by BenOnTrack** [reply]
I will try to reproduce the issue. In the meantime, when you mention it worked on your end, 
for the NE do you use:https://www.naturalearthdata.com/downloads/10m-cultural-vectors/10m-admin-0-countries/
And for the .process file do you have
```
function attribute_function(attr,layer)
	return attr
end
```
or
```
function attribute_function(attr,layer)
end
```
I've also changed the .config to
`"source_columns": ["ADMIN"]`
to remove un-necessary data.

Anyway I'll re-try the entire process and will post all full files.

**Reply by cldellow** [reply]
Yes, that's the NE data I'm using.

My config files are at: https://gist.github.com/cldellow/1483395f32d34ae2454148d9e4fb9988

Full disclosure: I haven't tried to author a style file that renders the labels, I'm relying on the fact that pmtiles.io shows the ADMIN/ISO_A2/ISO_A3 data when looking at the pmtiles file.

**Reply by BenOnTrack** [reply]
Thank you. 
But nvm, tl;dr I've found the issue - my style post-processing was slightly too fancy and only works with {name} id, hence discarding the initial {ADMIN}! 

thanks for the support :) 
I'm keeping this thread open for the LayerAsCentroid proposal for the attribute

**Reply by cldellow** [reply]
phew, good to hear!

**Comment by BenOnTrack** [comment]
The remaining issue though is the way it's logged within the tiles - it looks like it's a polygon text rather than a node. So i'm stil convinced there's a better way to do ?

 ![Screen Shot 2024-09-11 at 10 18 34 PM](https://github.com/user-attachments/assets/8505b1f6-c870-45a2-b2cf-d79ed53855f3)

**Reply by cldellow** [reply]
tilemaker's support for shapefiles/GeoJSON passes the underlying shape into the output pmtiles/mbtiles as-is.

So the NE data presumably has 1 multipolygon for Japan, which gets emitted as such. Most map renderers do a poor job labelling polygons (see, e.g. https://github.com/mapbox/mapbox-gl-js/issues/7846)

The workaround is to explicitly produce a point to use as the anchor for the label. You'll have to use some other tool to transform the shape to achieve that: tilemaker's `LayerAsCentroid` function isn't available in the context of `attribute_function`.

That might be a useful extension, though. I wonder if @systemed has thoughts?

It looks like there's already a magic meta key called `_minzoom`: https://github.com/systemed/tilemaker/blob/eab08d189ad97ddf5db7d915bcabe42ad3dab6af/docs/CONFIGURATION.md?plain=1#L212

Maybe there could be another magic meta key called `_centroid` which, if truthy, emits a point using the polylabel algorithm.

**Reply by BenOnTrack** [reply]
I was exactly thinking about that too :-)
